### PR TITLE
feat: implemented edit list page

### LINKED ISSUE #1977: Feature: Edit List Page
### Type of feature

🍕 Feature

### Current behavior

Users can't currently edit Lists

### [Figma file link](https://www.figma.com/file/OpVX6WT7dmWqnwRuEvADMF/Design-Lab?type=design&node-id=11428-10752&mode=design&t=N96wMixKQXsE4uUR-4)

**Parent Issue:**

### New Page - OPEN TO FEEDBACK
The idea is for users to be able to edit lists through a new "Edit list" page (we can reuse parts of the structure of the "Edit Insights Page" page).

This new page should allow users to:
- Edit the list name
- Edit page visibility
- Add new contributors
- See and remove existing contributors

image

### Removing Contributors
Users should be able to see a list of the contributors on the list, with the option to filter the list, and remove contributors.
When deleting a user, users should see a toast (reuse existing toast component) indicating that a user has been removed, and **with the option to undo the action and add the contributor back to the list (important!!)** .

Notes on usabilty:

image

### Adding contributors
The "add new contributors" button will take users to a new page, similar to the "explore contributors" page, here's the issue: #1978 

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by nickytonline**
## Description

Implements the edit list page including an edit list button available on the list's page in the hub.

~Note that search currently returns the same list as the page of data currently displayed. https://github.com/open-sauced/api/pull/405 fixes that for the `/v1/lists/{id}/contributors` endpoint so it's searchable like `/v1/lists/contributors` via a contributor search term. The rest all works.~ Merged and deployed.

Demo 👉 https://www.loom.com/share/bbe9a820de45438489419e31a76272e7?sid=7d7545f6-2d98-48ff-8563-24197d17c4bf

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Fixes #1977

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by OgDev-01** [comment]
Great stuff, @nickytonline 👏👏. 

Some observations with the pagination... if i next the page using the arrow icon, it doesn’t reflect on the page number.
https://github.com/open-sauced/app/assets/62995161/6e4240b6-cf25-46fa-a334-2124ef5f7082


And secondly, if we have up to 100 pages, will this truncate after XYZ length?

**Comment by nickytonline** [comment]
> Great stuff, @nickytonline 👏👏.
> 
> Some observations with the pagination... if i next the page using the arrow icon, it doesn’t reflect on the page number. [open-sauced/app/assets/62995161/6e4240b6-cf25-46fa-a334-2124ef5f7082](https://github.com/open-sauced/app/assets/62995161/6e4240b6-cf25-46fa-a334-2124ef5f7082)
> 
> And secondly, if we have up to 100 pages, will this truncate after XYZ length?

Thanks for reviewing @OgDev-01! I hadn't thought of that as I had assumed the paging component managed all this internally, but I hadn't checked.

I might just go to our default paging component for now since a lot of that is baked in.

**Review by jpmcb — COMMENTED** [review]
The copy in this poppup looks to not be centered - not sure where else this gets used, not seeing it else where in the app

Screenshot 2023-11-07 at 1 21 29 PM

---

The user experience of selecting "edit list" -> "Add new contributors" -> adding contirbutors in the list -> the above popup -> selecting "Done" -> being back to the list overview page is abit confusing.

My expectation would be that after selecting "Done" I'd be taken back to the list edit page with those new contributors populated. If you think of it like a stack, popping off the stack all the way out to the list overview page feels like one too many.

---

Looks like this list page needs some styling? Note the off white background:

Screenshot 2023-11-07 at 1 32 54 PM

---

This is overall looking amazing!!! Nice work Nick!

**Comment by nickytonline** [comment]
The API has been updated, so you can now filter on user list contributors.

Just sorting out the styling of the paging @OgDev-01.

**Comment by nickytonline** [comment]
> The copy in this poppup looks to not be centered - not sure where else this gets used, not seeing it else where in the app
> 
> Screenshot 2023-11-07 at 1 21 29 PM
> The user experience of selecting "edit list" -> "Add new contributors" -> adding contirbutors in the list -> the above popup -> selecting "Done" -> being back to the list overview page is abit confusing.
> 
> My expectation would be that after selecting "Done" I'd be taken back to the list edit page with those new contributors populated. If you think of it like a stack, popping off the stack all the way out to the list overview page feels like one too many.
> 
> Looks like this list page needs some styling? Note the off white background:
> 
> Screenshot 2023-11-07 at 1 32 54 PM
> This is overall looking amazing!!! Nice work Nick!

I think the centering issue is just on Firefox as it's fine in Chromium based browsers. Gonna check in Safari as well.

I believe the original designs had you going back to the list overview, but I agree going back to the edit page makes more sense.

**Comment by nickytonline** [comment]
> > The copy in this poppup looks to not be centered - not sure where else this gets used, not seeing it else where in the app
> > Screenshot 2023-11-07 at 1 21 29 PM
> > The user experience of selecting "edit list" -> "Add new contributors" -> adding contirbutors in the list -> the above popup -> selecting "Done" -> being back to the list overview page is abit confusing.
> > My expectation would be that after selecting "Done" I'd be taken back to the list edit page with those new contributors populated. If you think of it like a stack, popping off the stack all the way out to the list overview page feels like one too many.
> > Looks like this list page needs some styling? Note the off white background:
> > Screenshot 2023-11-07 at 1 32 54 PM
> > This is overall looking amazing!!! Nice work Nick!
> 
> I think the centering issue is just on Firefox as it's fine in Chromium based browsers. Gonna check in Safari as well.
> 
> I believe the original designs had you going back to the list overview, but I agree going back to the edit page makes more sense.

All fixed @jpmcb

**Review by brandonroberts — APPROVED** [review]
Works great from my testing! 🚢

**Comment by nickytonline** [comment]
Still sorting out the paging component since it's never been used with the `pages` prop aside from an empty array.

If you could take the rest for a spin again when you have a sec @OgDev-01, 🙏

**Comment by nickytonline** [comment]
@OgDev-01, I've fixed the paging logic so this is good to go whenever you have time for more feedback.


https://github.com/open-sauced/app/assets/833231/c356b507-3dee-4a2b-9c81-4e96e26c1b5c

**Review by jpmcb — APPROVED** [review]
Nice job!!! 👏🏼👏🏼👏🏼

File: components/molecules/Pagination/pagination.tsx [hunk 1]
```diff
@@ -43,42 +45,63 @@ const Pagination = ({
     onPageChange(page - 1);
   };
 
+  const middlePages = Array.from({ length: Math.min(pages.length, pageSize) }, (_, index) =>
+    selected <= Math.floor(pageSize / 2)
```
Comment (NsdHSO) [inline components/molecules/Pagination/pagination.tsx:49]: IMHO it is very hard to read this function, I tend to believe it must be refactored.

File: components/molecules/Pagination/pagination.tsx [hunk 2]
```diff
@@ -43,42 +45,63 @@ const Pagination = ({
...
             className="text-light-slate-9 disabled:text-light-slate-7"
             disabled={!hasNextPage ? true : false}
-            onClick={() => handleNext()}
+            onClick={() => {
+              if (!hasNextPage) {
+                return;
+              }
+
+              handleNext();
+              debugger;
```
Comment (NsdHSO) [inline components/molecules/Pagination/pagination.tsx:102]:
```suggestion
```

Comment (nickytonline) [reply components/molecules/Pagination/pagination.tsx:102]: 😬 We should a lint rule for that.

Comment (nickytonline) [reply components/molecules/Pagination/pagination.tsx:102]: I created #2082 for that.

Comment (NsdHSO) [reply components/molecules/Pagination/pagination.tsx]: Tnx 💯

Comment (nickytonline) [reply components/molecules/Pagination/pagination.tsx]: PR is up for the `no-debugger` ESLint rule. 😅 #2083

File: components/molecules/Pagination/pagination.tsx (hunk 1 above)
Comment (nickytonline) [reply components/molecules/Pagination/pagination.tsx:49]: I'll go ahead and put a comment in the code about it. We're probably refactoring that whole paging mechanism. This feature is the first feature to use the paging in that mode.

**Review by OgDev-01 — APPROVED** [review]
Bravo! 🍕

**Comment by nickytonline** [comment]
I chatted with @isabensusan and she made a couple style changes. I'm going to go ahead and merge this now that it's got her seal of approval. ✅

**Pull request merged**